### ISSUE TITLE
As a developer, I want to write a wiki for our users, using the VLab wiki as a starting point.

### ISSUE THREAD

**Issue by HankHerr-NOAA** [NOAA-OWP/wres#20]
We need a wiki for our user explaining how to obtain and run the software, how to declare an evaluation, and other aspects which are TBD. My plan is to start with a skeleton for an entry level page, then add the Declaration Language wiki from VLab one level down, with other wikis that are referred to migrated as well. I'll edit all wikis to make sure that they do not refer to NWC specific services or files.  Once the Declaration Language wiki is migrated, we'll add other pages as appropriate.  

I'm going to put the pages in the wiki, directly, but, eventually, we'll probably want to source control them. We can discuss that later.

Hank

**Comment by HankHerr-NOAA** [comment]
The page I'm going to use to translate from VLab Redmine textile format to GitHub markdown is here:

https://pandoc.org/try/

The **from** format is textile, and the **to** format is markdown_github. But, first, I need to create a top level minimal wiki.

Hank

**Comment by HankHerr-NOAA** [comment]
Turns out that the initial migration will point to the wiki pages from VLab. For example:

"[Units of measurement](https://vlab.OMITTED)"

I just need to find the VLab URLs, follow them, migrate the pointed to wiki, and then modify the VLab URLs appropriately.

Note that it appears the converting tool adds new lines inside of paragraphs enforcing a column-width limit on the text. However, when the wiki is viewed via browser, those new-lines are removed and the column width is controlled by the browser. If you want to see what I'm talking about, view the Declaration Language wiki, which I just created, and note the column width, then edit it and note that the column width is different.

**Comment by HankHerr-NOAA** [comment]
I'm about to convert the Units of Measurement wiki from VLab to GitHub. However, I need to add a task for later, but I don't see an ability to add a checklist item. Its probably because I'm overlooking it. The following task must be completed later:

**Add the footer in the Units of Measurement VLab wiki to the new Units of Measurement GitHub wiki and figure out how to properly do super scripts.**

I'm going to save the Units of Measurement wiki as I have it now and return to it later.

Hank

**Comment by HankHerr-NOAA** [comment]
There is a reference to a Google Drive document in the declaration language wiki:

The list of supported metrics is contained in this spreadsheet: [list of
metrics supported by
WRES](https://drive.google.com/open?id=1aRUvdktBdD0ZRMmnp4xT6P3BqmGMJGX8qCtMbGuf_r0).

That will need to be open to the public or we'll need to move that to a wiki. **Which is better?**

**Comment by HankHerr-NOAA** [comment]
For the Units of Measurement task in bold, above, I can see the links defined in the wiki page when I edit it, but they do not show up in the text when viewed/previewed. However, the links do appear to be working.  I'll have to investigate how to better handle those footnotes.

Hank

**Comment by HankHerr-NOAA** [comment]
I fixed the units of measurement links. They now appears as super scripts and there is a list at the bottom.

**Comment by epag** [comment]
> I'm about to convert the Units of Measurement wiki from VLab to GitHub. However, I need to add a task for later, but I don't see an ability to add a checklist item. Its probably because I'm overlooking it. The following task must be completed later:

Tickets support markdown so you can add a tasklist 
![image](https://github.com/NOAA-OWP/wres/assets/7482743/5d884c00-61d1-4778-b92e-ff0496277efc)

So you can edit the initial issue to contain a checklist 

- [ ] You can also convert these into sub issues

**Comment by HankHerr-NOAA** [comment]
The tool I'm using (noted above) does not appear to handle code tags well. I just manually went through the Declaration Language wiki modifying the code tags to use the appropriate notation discussed in the meeting. I still have to modify the other wikis, but they aren't nearly as large.

**Comment by HankHerr-NOAA** [comment]
The other code tag were fixed. However, in the Units of Measurement, I did fix the superscripts to appear as super scripts, and the list of references at the bottom looks good, but the references to those super scripts do not. Specifically, I see this:

blah[<sup>1</sup>]

I need to super script to just the number, not brackets around it, and I need it to link to the reference at the end. I'm not sure if that's possible. 

There are other mistakes in the units of measurement wiki that I'm fixing now.

Hank

**Comment by HankHerr-NOAA** [comment]
Looks like the Unit of Measurement wiki was never updated to YAML. I'm going to try to do that after I check my emails,

Hank

**Comment by HankHerr-NOAA** [comment]
Done. There are still probably mistakes, but this is all I can do for now.

Hank

**Comment by HankHerr-NOAA** [comment]
@james-d-brown: Is the Declaration Language wiki supposed to define every declaration option at some point, even if its a brief mention in a table? I ask because `resale_lenience` is not currently described in that wiki, though it is mentioned in the time scale wiki which is linked.  

Thanks,

Hank

P.S. We'll see if the '@' results in an email being sent to you.

**Comment by HankHerr-NOAA** [comment]
I'm going to go through the list of every wiki we offer in the VLab WRES User Support project. For each one, I'm going to check if its a reasonable candidate for migration to GitHub. My next comment will provide a list of those candidates.  

Candidates will be rejected if its too NWC-centric or if its part of the Cookbook, which is GUI-centric. 

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
And... never mind. For now.

I realized that I did not properly migrate all of the wiki linked to from within the Declaration Language. There are at least a couple remaining that were referenced using [[...]] instead of URL:

Using covariates as filters
Evaluation summary statistics

I want to migrate those, first.  Having looked through the overall list of wikis, there are quite a few other candidates, as well, including a wiki on "Pooling geographic features" (which is surprisingly,to me, not referenced in Declaration Language), the CSV output format, format requirements for netCDF, and others. It might be best to just spend 15 minutes going through the list during a meeting and deciding which to migrate.

Let me do the two linked from within Declaration Language that I overlooked. 

Hank

**Comment by HankHerr-NOAA** [comment]
I added a few more wikis and cleaned up the indentation of code examples in Declaration Language. One that I don't have time to do today is the examples wiki, Complete Examples of Evaluation Declarations. Thats referred to from within Section 18 of the declaration language wiki. I also want to double check how I'm linking to the other wikis, which currently involves treating it as any external URL, with the format being **[display text](URL to link)**.  I discovered shorthand for displaying images that uses double-brackets. Something similar probably exists for linking other pages within the same wiki.

Also, to add images, I had to clone the wiki repository, add the images to the repo, and then commit/push. Once I entered my username and password, it pushed fine. I found now way to upload images via the browser interface. What I'm not sure of is if the wiki, itself, is open to the public as with the main repository, and if branch protections apply. In other words, was I able to push without issue because I was one of the maintainers/admin? Or because the repository doesn't have branch protections requiring pull requests? If its the latter, does that mean anyone can push to the wiki repository?  

Known tasks for next week:

- Clean up indentation for code examples in all wiki pages.
- Complete Examples of Evaluation Declaration wikis.
- Any other wikis directly or indirectly referenced from within the wikis I've already migrated.

More next week. Have a great weekend!

Hank

**Comment by epag** [comment]
Do you have a page that I can look at that shows this as a good example? I think it would be better to accept this short coming rather than do everything manually at least

**Comment by HankHerr-NOAA** [comment]
You can see it in any of the wiki pages I created through migration from textile. For example: Declaration Language:

https://github.com/NOAA-OWP/wres/wiki/Declaration-Language

Just click Edit and you can see the column width limit applied.

Hank

**Comment by epag** [comment]
I think that looks fine personally, especially if the alternative is not using the tool or manually making changes

**Comment by HankHerr-NOAA** [comment]
The alternative would probably be me removing the excess new-lines by hand. Although it may be possible to script it in a reasonable way.  The change would be made once per page, so it wouldn't be too torturous. I'm just wondering if, conceptually, if a column limit is sensible.

Hank

**Comment by james-d-brown** [comment]
If this is just a dev weakness, we can probably live with it. If it affects the product seen by users, probably not.

**Comment by HankHerr-NOAA** [comment]
I've seen no wiki impacted by the column width limit as of yet. I guess that means we can live with it. Just be ready when you go in to modify a wiki to see a column width that may be kind of ugly to work with.

I had said I should return to the wiki changes tomorrow, but I doubt that will happen. Probably won't be until Friday at this point. Repeating here the known tasks that remain to be done:

- Clean up indentation for code examples in all wiki pages.
- Complete Examples of Evaluation Declaration wikis.
- Any other wikis directly or indirectly referenced from within the wikis I've already migrated.

Hank

**Comment by james-d-brown** [comment]
Other issues:

- TOC needed fairly urgently for any big wikis as, for example, the declaration language wiki will be very hard to use without one;
- In the declaration language wiki, I see this, which will not be publicly available, I think: "The list of supported metrics is contained in this spreadsheet: [list of
metrics supported by
WRES](https://drive.google.com/open?id=1aRUvdktBdD0ZRMmnp4xT6P3BqmGMJGX8qCtMbGuf_r0)."

**Comment by james-d-brown** [comment]
Priority item: all of the tabulated examples in the declaration language wiki are missing. Not good insofar as we will have internal users looking at our github docs now. Aside: I cannot update the wiki for variable aliases until the tabulated examples are in place...

**Comment by HankHerr-NOAA** [comment]
Let me take a look at the tabulated examples.

I commented in the VLab ticket on the ToC being on the right hand side. I'm assuming you saw this and decided it was insufficient?

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
I see what you mean about the "Examples in context". Checking to see if the text from the wiki was migrated but is not showing up, or if it just wasn't migrated at all.

Hank

**Comment by HankHerr-NOAA** [comment]
Not migrated. I wonder why?  

I guess I need to try to migrate each individually. This will fun,

Hank

**Comment by HankHerr-NOAA** [comment]
I see why it wasn't migrated. The multiline example does not come through well in the wiki. I'm going to see if I can turn the examples into single-line examples and use a new-line character to have it display as multiple lines when viewed.

**Comment by HankHerr-NOAA** [comment]
Using embedded HTML can at least get a multiline example to show up:

| `sources` | To clarify the list of sources to evaluate when other options are present for the dataset as a whole. | <html>observed:<br>  sources: some_observations.csv<br>  variable: streamflow<br>predicted: some_predictions.csv</html> |

My problem is that I don't know HTML. For example, why isn't it respecting the whitespace after the <br>s?  It appears to get rid of starting and ending whitespace.

More work to do,

Hank

**Comment by HankHerr-NOAA** [comment]
Here is what I put in for the example in my previous comment:

```<html>observed:<br>  sources: some_observations.csv<br>  variable: streamflow<br>predicted: some_predictions.csv</html>```

Hank

**Comment by HankHerr-NOAA** [comment]
I think I have a solution. This is the text that is in the Example in context column:

```
<pre><code class="yaml">
observed:
  sources: some_observations.csv
  variable: streamflow
predicted: some_predictions.csv
</code></pre
```

Here is an example that works, keeping the pre-tags and using HTML breaks, ```<br>```, instead of new lines. This is the raw text:

```
| `sources` | To clarify the list of sources to evaluate when other options are present for the dataset as a whole. | <pre>observed:<br>  sources: some_observations.csv<br>  variable: streamflow<br>predicted: some_predictions.csv</pre> |
```

This what it looks like when viewing (for some reason, the code example wraps to the next line when previewing, but its on the same line in the wiki):

| `sources` | To clarify the list of sources to evaluate when other options are present for the dataset as a whole. | <pre>observed:<br>  sources: some_observations.csv<br>  variable: streamflow<br>predicted: some_predictions.csv</pre> |

Good enough? I'll see if I can figure out how to get it to use color,

Hank

**Comment by epag** [comment]
The alternative if you want to use color it seems is to create the table in html which will allow the use of a code block in the table:

https://stackoverflow.com/questions/28508141/code-block-inside-table-row-in-markdown

**Comment by HankHerr-NOAA** [comment]
Yes, I saw that. And it horrified me. :)

I started migrating using the <pre> and <br> method in the first table of the Declaration Language wiki and made it a bit more than halfway through the table:

https://github.com/NOAA-OWP/wres/wiki/Declaration-Language#34-how-do-i-clarify-the-datasets-to-evaluate-such-as-the-variable-to-use

You can go into the editor to see what it looks like when editing it. Its a painful process that feels like I'm being tortured, particularly when Is see how many cells need to be updated. In that table, there was also some multi-line cells in the the second column that were incomplete because it would only migrated the first line.  I fixed those, as well.  

**Please compare what is currently in that table with what Evan shared in the link and let me know which path forward is preferred.** Either is a lot of work, but the HTML solution looks like it would be more. That is, unless a script of some sort can be created to migrate it.  I may look into that after lunch.

Thanks,

Hank

**Comment by james-d-brown** [comment]
Agree, that is pretty horrendous. It takes a lot more effort to get to something that is worse than our current wiki. Sigh.

**Comment by james-d-brown** [comment]
The first step is just to make it usable, which is what you have done, Hank. Adding color will have to wait. We need to get this done for the v6.24 deployment so that it can be augmented with info about variable aliases.

**Comment by HankHerr-NOAA** [comment]
I created a process that can use a basic editor and find-and-replace to modify table content starting with the first row of actual content. That is, don't put the column header row and row of dashes through the process I'm about to describe.  Those two rows are migrated correctly, anyway.

So, starting with the first row of content from the VLab wiki, do the following:

- replace "@" with "```"
- replace "\n-" with "\<br\>-" (use the editor to copy the newline and hyphen and paste it into the from field of the replace; that works with nedit). This allows for hyphened lists in the VLab wiki to be seen in the GitHub wiki.
- replace "\<code class="yaml"\>"  with nothing
- replace "\</code\>" with nothing
- Manually go through each \<pre\> example and swap out "\n" for "\<br\>" (again, copy the new-line from the text and paste it into the from field), making sure the "\<pre\>" itself is not followed by "\<br\>" (which causes an empty space in the code snippet).  This is the hard step, but **nedit** has this nice ctrl-t shortcut that makes it pretty easy.

The result of this process can now be seen in the first table of the wiki. The steps above were designed for that first table. There are likely other steps that may be needed if other tables don't follow the same pattern. I'll address those when I find them.

I still need to eat lunch. so I'll do that now. I'll then try to fix the tables in the Declaration Language wiki after lunch, unless there is something else in my email list that takes priority.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Turning my attention to other wikis, I'm now going to do this:

- Clean up indentation for code examples in all wiki pages.

Hank

**Comment by HankHerr-NOAA** [comment]
That's done.  That leaves just adding the missing wiki, such as the examples wiki, which links to five separate example sub-wikis.  Let me see how the migration website handles the top-level wiki,

Hank

**Comment by HankHerr-NOAA** [comment]
For the examples, I have created the first draft of the wiki, but note that the instructions for executing the examples needs to be updated. Specifically, the data backing examples 1 - 3 are not available via GitHub, and example 5 only works for COWRES users. I did update example 5 to omit the WRDS URLs. 

Hank

**Comment by HankHerr-NOAA** [comment]
There are two placeholders for how to execute the examples: one for using COWRES and one for local execution. For location execution, as mentioned above, the data for examples 1-3 are not available via GitHub.  For COWRES, the question is, do we want to include execution examples in the GitHub wiki, and, if so, is it safe to share the directory where the data is available for those evaluations?  

I don't think I'm going to address this today; perhaps on Friday.  Also on Friday, I need to look through the wiki, again, for other VLab wikis that are referred to but have not been migrated.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
@james-d-brown , @epag ... Looking at the examples wiki, https://github.com/NOAA-OWP/wres/wiki/Complete-Examples-of-Evaluation-Declarations, I can make the "complete examples" runnable by either pointing them to the latest system testing tarball in Google Drive (opened to the public) or if there is a GitHub location that can host the 328 MB tarball, that might be better.  The tarball only stores a small part of the overall system testing data, that's why its only 328 MBs, but that happens to be enough to cover examples 1 - 3.  **Are we okay with sharing the system testing data? Is there a better location than Google Drive with the file opened up to the public (read-only)?**

Example 4 uses the smalldata directory in the repository.

Example 5 is only intended for users with access to WRDS services.  

Thanks,

Hank

**Comment by james-d-brown** [comment]
Yes, we should share the data. There is almost certainly a better location than g-drive. Whether there is a better, equally free, location, idk. There are free data repos for scientific research, such as Zenodo, which is EU-centric. There may be something similar in the US, for example. You should probably check with OWP whether sharing data with public read-only on gdrive is acceptable.

**Comment by james-d-brown** [comment]
This problem will again arise when hosting the system test datasets for our own downstream system tests, which will be necessary to migrate them into the github workflow so that they are visible to others contributing to the codebase in future. It is the same problem, and the solution can be the same.

**Comment by HankHerr-NOAA** [comment]
If we could turn the tarball into something equivalent to a release artifact (from GitHub's perspective), that could work:

[Distributing large binaries](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-large-files-on-github#distributing-large-binaries)

It probably would be cleaner to find a true, large file solution, however.

Hank

**Comment by HankHerr-NOAA** [comment]
There is also GitHub packages:

https://github.com/NOAA-OWP/wres/packages

That seems to require selection of a registry-type, so I'm not sure its really appropriate for just uploading/downloading large files.  Probably not.

Hank

**Comment by HankHerr-NOAA** [comment]
As for the rules of Google Drive, I'll reach out to folks to see if anyone knows if we are allowed to open up Google Drive files to the public as read-only.

Hank

**Comment by HankHerr-NOAA** [comment]
I've emailed Jeff and cc'd you all asking about sharing a file on Google Drive.

In the mean time, Evan, if you know of a good GitHub solution, let us know. Thanks,

Hank

**Comment by james-d-brown** [comment]
Thanks. I think the main problem with the g-drive solution is that will be temporary, because we'll need to solve the same problem for the system tests in a way that is accessible to a github pipeline.

**Comment by HankHerr-NOAA** [comment]
Put the data in a WRES repo "submodule"?

https://github.com/orgs/community/discussions/25527

I think that solution might work for the 300 MB tarball, with each file in that tarball being separately gzipped and stored in the repo, but I doubt that would be acceptable for the gridded system testing data which is much larger. But perhaps that can wait until we have time to shrink the system testing data.  

Hank

**Comment by james-d-brown** [comment]
Perhaps mechanically, but it isn't very purist, sticking blobs in a source code repo.

**Comment by HankHerr-NOAA** [comment]
Investigating this:

In the declaration language wiki, I see this, which will not be publicly available, I think: "The list of supported metrics is contained in this spreadsheet: list of metrics supported by WRES."

Hank

**Comment by epag** [comment]
If we can get approved to have a public facing gdrive location than we can just download all of our testing information from there, that may be the best/simplist approach

**Comment by HankHerr-NOAA** [comment]
My first attempt at a list of metrics available uses HTML converted from the Google Sheet:

https://github.com/NOAA-OWP/wres/wiki/List-of-Metrics-Available

I think we should be able to edit that HTML directly; it doesn't look too painful. Thoughts? It is almost certainly possible to better format the table, but my main goal for now is just to migrate the content to GitHub.

Hank

**Comment by james-d-brown** [comment]
Sure it will ultimately work as a solution, but would need work. The cell references are alphanumeric like "C1", so would need column labels in the first table for example and row numbers fixed.

**Comment by HankHerr-NOAA** [comment]
I had overlooked the row numbers and had actually removed the column header line. I just brought that back in and fixed the row numbers,

Hank

**Comment by HankHerr-NOAA** [comment]
Oh, and FYI...

You hand mentioned putting the markdowns in the repo. Just so you know, they are currently already in a repository, and, in fact, I had to use that repository to upload the images in the sampling uncertainty wiki. This is the repo:

https://github.com/NOAA-OWP/wres.wiki.git

One thing I need to do is make sure that that repo is read-only to everyone but maintainers.  

Hank

**Comment by HankHerr-NOAA** [comment]
We have checked these options checked in the repository settings;

Restrict editing to users in teams with push access only Loading
Public wikis will still be readable by everyone.

So we should be good,

Hank

**Comment by james-d-brown** [comment]
Sounds good. I made a few edits to the table. The first table is still a bit clunky, TBH, and could be improved (not super clear what are the row vs column headings here), but it is more effort than I am currently willing to invest.

**Comment by james-d-brown** [comment]
( OK, a couple more edits, but that is it. )

**Comment by HankHerr-NOAA** [comment]
Returning to this... I"m now going to add disclaimers to the top of migrated VLab wikis. I'm thinking about this:

**This wiki has been migrated to GitHub here "here":[link]. Please visit that wiki for the latest version. This page will no longer be modified and has been deprecated for removal in the future.**

Sound good?

Hank

**Comment by HankHerr-NOAA** [comment]
I'm also adding "(DEPRECATED)" to the top level headers in the wiki pages.  To see an example, go to the Complete Examples of Evaluation Declaration wiki in VLab. 

Hank

**Comment by HankHerr-NOAA** [comment]
I added the disclaimer to all VLab wikis that have been migrated.  Note that its not a big deal to change the text, so, if you have any recommended wording change to the disclaimer, please let me know.  

I'm going to try to find any remaining wiki to migrated by combing through the Declaration Language to see if there are any other directly or indirectly referenced that I failed to migrate.

Hank

**Comment by james-d-brown** [comment]
Sounds good.

**Comment by HankHerr-NOAA** [comment]
Looking through the VLab wiki for other wikis that may need to be fully or partially migrated:

- The **Instructions for using WRES**, which includes a general look at how to perform evaluations, but is very COWRES-centric.
- **Configuring the Raw NWM Data Source**, which is generally applicable to downloaded NWM netCDF files, but includes references to d-store. It also includes references to Nomads and Amazon S3, but I'm not sure those still work.
- **Format Requirements for NetCDF Files**
- **Obtaining_and_using_the_WRES_as_a_standalone_application**, which is a more detailed version of the Readme, I think.
- **Output Format Description for CSV2**
- **Pooling geographic features**. May need to look into why that wiki is not referenced in the Declaration Language wiki.
- **Introductory_Resources_on_Forecast_Verification**
- **WRES_Geographic_Location_Table** maybe. Its a reproduction of the WRDS crosswalk in table form that is probably out of date. If we migrate it, we should update it.
- **WRES Local Server**

The other wikis are either GUI centric, like the Cookbook, or geared toward the COWRES or WRDS. If and when we migrate wikis, above, we should identify a reasonable way to link them to either the top level wiki, either directly or via another wiki such as Declaration Language.

Hank

**Comment by HankHerr-NOAA** [comment]
Looking at **Pooling geographic features**, I think it belongs as a new Section 9.3 in the current Section 9, "How do I declare pools of data that should be evaluated separately?". That section is focused entirely on temporal pooling currently, but the title of the section indicates pooling in general. so it makes sense to cover feature pooling in 9.3, I think. 

The wiki, which would be referenced in Section 9.3, includes mention of WRDS, so it will require some editing to remove exact URLs and rephrase things if needed.

Working on that wiki now,

Hank

**Comment by HankHerr-NOAA** [comment]
I added the Pooling geographic features wiki and linked it in a new Section 9.3 of the Declaration Language wiki. 

For this migration, I turned off the wrapping feature of the migration tool to see how much of a difference it made. It significantly easier to edit a wiki that has the wrapping turned off, relying on the built in GitHub wiki editor to wrap the text on screen. I think, moving forward, I'm going to continue to not wrap the text, and then fix the wikis I've already migrated by hand as time allows. The Declaration Language wiki will be a problem, because of how long it is, but the rest are relative short and shouldn't take too long to fix on an as needed basis (i.e., whenever I go in to make a change).

I'm going to note in my earlier comment that the Pooling geographic features wiki was migrated.  

Hank

**Comment by james-d-brown** [comment]
Hank, I don't think that is the right place for it. There is a danger the declaration language wiki becomes a vast general-purpose monolith of everything the wres can do, which is not the purpose. The purpose is specifically to describe the declaration options and what they do. File formats have nothing to do with declaration because the formats are auto-detected. We should include separate wiki pages for separate features or attributes of wres and keep the declaration language wiki exclusively focused on declaration options.

**Comment by james-d-brown** [comment]
The closest thing to something useful in that area, and consistent with the purpose of the wiki, would be to enumerate the `interface` options associated with a `source`.

**Comment by james-d-brown** [comment]
> The closest thing to something useful in that area, and consistent with the purpose of the wiki, would be to enumerate the `interface` options associated with a `source`.

Specifically, that is an omission from the table in section 3.4, for which a row should descibe the `interface` and list the options, then provide an example in context.

**Comment by HankHerr-NOAA** [comment]
When you say, "declaration language wiki", lower case, are you referring to the Declaration Language wiki, specifically, or the entire portion of the wiki connected to declaration language? In other words, if I moved the file formats section 3.5 to a separate wiki and referenced it appropriately, would that be sufficient?

The declaration language wiki is already very big and I noted you did not include the file format list, I just wasn't sure of the reason.  

Thanks,  

Hank

**Comment by HankHerr-NOAA** [comment]
Okay. Migrating the instructions wiki may need to be next. I've been pushing it off, because of the COWRES dependencies, but you're right that a front door is needed.  

I'll remove 3.5 from the wiki and then start working on the instructions, which is going to take some time.

Hank

**Comment by HankHerr-NOAA** [comment]
I'm not sure how to mark this, but I need to put this work on-hold. We had a breakthrough with the Podman work, so I'm moving to a VLab ticket.

Hank

**Comment by epag** [comment]
Ill add an on-hold status to our project

**Comment by HankHerr-NOAA** [comment]
I added https://github.com/NOAA-OWP/wres/wiki/Introductory-Resources-on-Forecast-Verification and https://github.com/NOAA-OWP/wres/wiki/WRES-Local-Server. The latter may need some further clean up.

I also added the instructions wiki, but did not complete it due to other activities. Still, I saved it so that I don't lose progress. I'm attempting to walk a fine line with that wiki, which was originally written specifically for the COWRES.

I'll try to wrap up the new documentation by this afternoon, but it may not be until tomorrow.

Hank

**Comment by HankHerr-NOAA** [comment]
James:

I asked this in the meeting, but you were having connection issues. I was thinking about adding direct links to the VLab WRES User Support wiki in the GitHub wiki, with something letting the reader know that only folks with access to NOAA VLab will be able to follow the links.  This would make it easier to redirect NWS readers of the GitHub wiki to VLab wikis that can fill in the gaps, providing specific URLs for NWC services.  Currently, I only generally redirect the user to "VLab project WRES User Support wiki" with out specific of where to go.  

Do you see any reason for concern in doing that?  Evan said he had no issues, but wondered if security folks might be concerned. He mentioned "security by obfuscation" as being a potential reason not to share the links, but the wiki is already readily findable using a simple web search.  Thoughts?

Hank

Hank

**Comment by james-d-brown** [comment]
I wouldn't create links to vlab, although not for security or policy reasons (you would need to address that to owp security/policy folks), rather because it's extremely clunky to swap between the two platforms. It would help to have some examples but if you mean WRDS, primarily, then I would describe the relevant features of WRDS in github wikis while making it clear that functionality is currently only accessible to NWS users (although, if WRDS is available on github then, in principle, they could instantiate something similar, albeit without the NWS data). In short, migrate all relevant wikis to github, obscuring specific URLs for NWCAL services - if they need those, put them on a vlab page and advertise it (but not via github - and make it clear where services are only accessible to NWS users.

Again, it may help to have an example of something that is awkward to translate (e.g., because it involves obscuring a lot more than simply the service URLs).

**Comment by HankHerr-NOAA** [comment]
You can look at what I'm translating now for an example:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Using-WRES

I don't want a separate set of instructions for COWRES users, when the overlap is pretty significant. Hence, I'd like one set of instructions that then points to COWRES- or WRDS-specific instructions in VLab where appropriate. Then again, when you look at the wiki, you may actually think it needs to be overhauled. It is pretty old, at this point, and perhaps instructions focused on the mechanics instead of the concepts would be more appropriate. 

Hank

**Comment by HankHerr-NOAA** [comment]
I'll return to this tomorrow. The instructions are only partially done, with some obvious formatting issues that need to be addressed (e.g., Step 1 is a different header level than the rest of the steps). I'll also need to address references related to COWRES, some of which (not all of which) are marked with TODOs.  

More tomorrow,

Hank

**Comment by james-d-brown** [comment]
I don't think there is any need for a separate set of instructions for COWRES. The COWRES is a cluster instance. Our github codebase supports that deployment. Thus, anything related to COWRES that is not couched in the http api more generally is probably a flawed wiki. As far as I can see, the only think that needs to be hidden is the URL to the instance we call COWRES.

**Comment by HankHerr-NOAA** [comment]
I'll take a look tomorrow. 

The idea of writing instructions for using the COWRES in GitHub, but not actually telling them the COWRES URL, seems a bit odd.  And, realistically, the only COWRES instance likely to be stood up any time soon, is the one that is already stood up at the NWC.  Its going to be difficult to talk about a COWRES without actually talking about _the_ COWRES.

Anyway, more tomorrow,

Hank

**Comment by james-d-brown** [comment]
It isn't odd. At least, it isn't odd compared to the alternative of splitting information between two separate platforms. In principle, the http api is something that external users will want to understand and, in some cases, implement. Again, if anything is referring to COWRES specifically, it is probably not framed correctly, rather it should be framed towards the http api for a cluster instance. The missing URL is a single tiny detail. No one needs a wiki for that, and vlab still exists for basic information about the status of the COWRES cluster, such as whether it is up or down.

**Comment by HankHerr-NOAA** [comment]
I'll take a crack at it today and see what I come up with.  

As for terminology, I was thinking about how to refer to a "COWRES" generally. Web service came to mind, but part of me is worried about potential confusion with "local server", since both are communicated with as if its a "web service", i.e., using something equivalent to **curl**. Its just that one is localhost and the other is available beyond that and allows for multiple workers.  As long as I'm careful with the terminology, it should be okay. Just flagging that as a worry.

Note that we have a whole host of wikis that will need to be written from scratch about a "web service" instance. We'll need a wiki explaining the design, components, and how to start it. Another talking about certificates, likely explaining how to create them (there are many). And potentially others. We have wikis, now, written for internal use, but they'll have to be modified to be appropriate for external use.

My goal today is to complete a draft of the top-level Instructions wiki. I also need to look over my migrations of the WRES Local Server and Introductory Resources wikis and make sure I've updated the VLab wiki with the this-has-been-migrated header.

Hank

**Comment by HankHerr-NOAA** [comment]
James:

A very minor question about style... I noticed that the wiki pages you originally authored tend to be named so that only the first letter is capitalized, like a sentence without a period.  The other wikis have followed title capitalization (nearly all words capitalized, except a, an, the, and other "short" words). Given the other wikis were there first, was there a reason why you chose to go that route?  If there was a good reason, then perhaps the other wikis should be retitled.

Thanks,

Hank

Hank

**Comment by HankHerr-NOAA** [comment]
The instructions wiki is ready for another review:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Using-WRES

There are a few referenced wikis that either need to be written or migrated. I'll get to them when I can.

Hank

**Comment by HankHerr-NOAA** [comment]
Done for today. I've migrated most if not all of the wikis that are downstream of the instructions wiki,

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Using-WRES

There is still a log to do. For example, we need to describe the WRES web-service so that developers will understand the concept and be able to use the files we provide. That may be quite the wiki. There are also various other development wikis that should probably be migrated, as well. We just need to be careful about what we share.

I'll return to this when I can, but, again, feel free to edit what I've already written as you see fit. I really don't want to "own" documentation. :)

Hank

**Comment by HankHerr-NOAA** [comment]
Reviewing what I have now to identify gaps. My intention is to resolve this ticket when I can and put it into UAT for 6.25, UAT consisting of reviews from you all. The wiki is essentially a mirror of what is in VLab for WRES User Support + some wikis from the WRES development project, all modified for GitHub. The biggest gaps I know of is the lack of a wiki explaining the WRES web-service approach and wikis for general development (not sure what we need there). 

I also have a gap where training material is linked to explain basic concepts, pointed to here:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Using-WRES#planning-the-evaluation

The slides I can readily turn into PDFs, but the training video is about 140 MBs in size. What is the upper limit I should apply to files share directly in the wiki (i.e. in the wres.wiki repository)? I'll need to look that up. I'll also need to determine if we are allowed to share training videos. Plus, that training video probably needs updating, perhaps via a Concepts, part 2 video or Advanced Concepts.  For example, we can talk about sampling uncertainty, covariates, summary statistics, and other features added since the 2020 training. We can tackle that in the future.

Reviewing the wikis now,

Hank

**Comment by HankHerr-NOAA** [comment]
@james-d-brown When you have a chance, can you please update the output CSV format wiki here:

https://github.com/NOAA-OWP/wres/wiki/Output-Format-Description-for-CSV2

It appears to at least be missing columns for the covariate. I'm not sure if other columns are missing. The example probably needs updating too.  

If you'd rather I create a separate ticket for that so that you can tackle it later on, just let me know. 

Thanks,

Hank

**Comment by james-d-brown** [comment]
Yeah, I can take a look tomorrow.

**Comment by HankHerr-NOAA** [comment]
I need to look over the Declaration Language linked wikis, but the others look okay now. I'm still referring to the WRES User Support VLab project in some wikis; we might decide to edit those to remove the references. I'll wait for a review.

I'm hoping to do a pass through for the Declaration Language wikis tomorrow, including reformatting them to remove the new-line line wrapping. As mentioned before, the editors available for the wiki work better when its continuous line wrapping instead of artificially inserted new-lines.  

More tomorrow,

Hank

**Comment by james-d-brown** [comment]
Yes, we should remove all references to vlab, these wikis need to be standalone.

**Comment by HankHerr-NOAA** [comment]
A typical example of where I refer to the WRES User Support VLab project is whenever I talk about WRDS. 

We need to mention WRDS as an option in the GitHub wiki, because we are pointing all users, RFC and OWP included, to the GitHub wiki, and they will need to use WRDS via COWRES. But, to use WRDS, those users have to know the URL, and that information will need to be provided in the WRES User Support wiki.  So how do I talk about WRDS without talking about the WRES User Support VLab project where the URL is available? 

Maybe if you had an example of how you would make the connection, that would help. Fortunately, there aren't too many wikis where I have to talk about WRDS, so making a change to the current approach would not take long, but I don't see a non-clunky solution to the problem.

Hank

**Comment by james-d-brown** [comment]
You can't talk about wikis that don't exist for all users. The best thing would be to inform NWS users separately where they can obtain NWS-centric information, such as service URLs. From the github wikis, the most you can do is to ask readers to contact the WRES team for NWS-centric information. We can then point them to the relevant NWS wikis (which should be advertised separately, as noted above, so this contact should not happen very often).

**Comment by HankHerr-NOAA** [comment]
"You can't talk about wikis that don't exist for all users."

I guess I'm not as hard lined on that as you are. If we are specific about what group of users a comment is directed towards (perhaps in a highlighted NOTE or something), then I'm not as worried about it. 

What I am worried about is pointing a group of users to a separate, external source of information without any connective tissue in the GitHub wiki (our primary wiki) explaining when that source of information needs to be referred to. Its going to put a burden on that group of users to make the connections themselves. To me, that's worse than including a link only some folks can follow (with a NOTE explaining why others won't be able to follow it). I'm particularly worried because the group of users we are talking about are our primary users: RFCs and OWP.

I need to catch a bus soon, but will give this some more thought tomorrow,

Hank

**Comment by james-d-brown** [comment]
Talking about wikis that don't exist for some users will be confusing for those users and there is, in any case, no possibility of providing a direct link, so it will be vague information. To do this properly, we will need to replace the user support vlab with a private github project for NWS users and then link the private wiki pages from the public ones, which will require the correct access privileges for the privileged information.

There is really no great way of solving this problem. Even the solution above is a compromise. However, if we are going to secure privileged information, it should at least be immediately accessible to those with the correct privileges. Mentioning wiki pages on some other platform that are accessible to a subset of users is not the best of a bad set of solutions.

We also need to think about the bigger picture here. When talking about external services, such as WRDS, it needs to be imagined in the context that other instances could exist in future, including non-NWS instances. It's the same when discussing the WRES-as-a-web-service. It would be a bad idea to develop a set of privileged wiki pages to describe COWRES when literally the only NWS-centric information about that service is the service URL. There could be other web service instances in future and, in general, service administrators would like to have some information about the web service api etc. without the baggage of a particular NWS instance as that baggage does not, in any way, help to understand the service.

**Comment by james-d-brown** [comment]
Added the covariate filters to the csv2 wiki. Also added the summary statistics columns, which were also missing, apparently.

**Comment by HankHerr-NOAA** [comment]
Using a private GitHub repo is, unfortunately, not a realistic option. The hoops that would have to be hopped through to get one visible to NWS folks are pretty severe (it can't be in the OWP GitHub organization, so the steps are different), and then every NWS user of the WRES would need to be explicitly added to the repo since it would be private.  

I think, to allow NWS users to make the connections more readily, it may be best to point out both wikis to RFC and OWP users, but then recommend they look at the WRES User Support wiki, first, which will then redirect them to the GitHub wiki. The VLab wiki front page will need to be changed to make sure WRDS, D-Store, and other NWC-centric capabilities are covered up front and in a way that's hard to miss.

I'm going to clean up the wikis to remove mentions of WRES User Support and make references to WRDS more generic. 

Hank

**Comment by james-d-brown** [comment]
Sounds good. If the NWS-centric information is that important/distinct for NWS users, then the NWS vlab wikis should be the front door for NWS users. Personally, I would be very surprised if anything other than a tiny amount of NWS-centric information were provided on the NWS wiki, with the rest linked to the front door here. It should probably be a one-pager.

**Comment by HankHerr-NOAA** [comment]
Yes, that's what I'm envisioning. The NWS-centric stuff includes WRDS (NWM, AHPS, observations, location service, thresholds), D-Store, and the COWRES URLs.  All of that will be up-front on the first page, but will need to link to some other wikis for details. For example, the current D-Store wiki includes example YAML for each NWM data type and a listing of versions and date ranges. I wouldn't want to put that in the up front wiki. Rather, I would identify the D-Store URL in the up front wiki and then link to the detailed wiki for more information.  Something like that.

Oh, and obviously the WRES GUI Cookbook documentation will continue to be in WRES User Support. That will need editing to refer to GitHub as appropriate instead of wikis that are deprecated.

Updating the VLab wiki will happen once I finish the next round of edits in the GitHub wiki... and can find the time (and willingness).

Hank

**Comment by HankHerr-NOAA** [comment]
James,

I'm editing the feature pooling wiki now and modified the content of "2. How does the WRES combine information from different geographic features?" to refer to summary statistics. Previously it said that "pooling statistics" was not available. Please see 

https://github.com/NOAA-OWP/wres/wiki/Pooling-geographic-features

and let me know if my change is correct.

In general, editing that wiki is a bit of a challenge. It makes significant references to WRDS, but you made an effort when you wrote it to specifically state that those portions of the wiki only apply to those with access to WRDS. I made some small edits, including omitting the WRDS URL and adding a NOTE for how to find it. I'm looking for another opinion on the changes I made, so please review it and let me know what you think.

Hank

Hank

**Comment by HankHerr-NOAA** [comment]
Thanks!

I think we need to rewrite the units wiki. Its very version specific, referring to before-and-after 5.14 in the first paragraph, which will make no sense to readers. That paragraph needs to be overhauled. The formatting of footnotes is off (its superscripted, but include [] around the numbers, which is ugly), which was probably a byproduct of the migration from texttile to markdown.  The style is also not the same as other wikis, which were mostly written by James or myself. There are other wikis written in different styles, as well, but this one just feels more jarring when visited. Perhaps its the whole version-dependency thing that is making it jarring.

Note that, with most other wikis, if it says something like, "As of release blah...", I just remove that phrase since I don't think its important with the move to GitHub. In the units wiki case, the references are are more embedded.

I'll put rewriting that wiki on my plate for tomorrow.

Hank

**Comment by HankHerr-NOAA** [comment]
I think I'm ready for the wiki to be reviewed, except for the units of measurement wiki which I'll edit later today or tomorrow. The known gaps, at this point, are as follows:

- A description of the WRES web-service. 
- Developer wikis, in general. 

I plan to write separate tickets for those wikis, allowing this ticket to be resolved for 6.25. I also need to update the VLab entry wiki appropriately, as well as the WRES Cookbook, but that will be tracked in VLab tickets.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Tomorrow, I plan to explore options for the side bar, hopefully allowing for organizing the wikis and making it easier to navigate between them.

For now, however, I'm hopping of early because of a headache. I'll work a bit extra tomorrow to make up the time. Have a great evening!

Hank

**Comment by HankHerr-NOAA** [comment]
I added a no-frills custom sidebar in the wiki:

https://github.com/NOAA-OWP/wres/wiki

The wiki sidebar is organized into two general types, instructions and declaration language, explained as follows:

- instructions and the wikis underneath describe the mechanics of how to run an evaluation and technical details of file formats (e.g., CSV and netCDF for input, CSV2 for output); and 
- declaration language and the wikis underneath describe the YAML (duh) and explain what the different declaration options do from a scientific/statistical point of view (e.g., sampling uncertainty, summary statistics, etc.)

The wikis underneath of each type are ordered alphabetically.

In addition to the above, at the bottom of the side bar are two other wikis referred to from the Home wiki, and, therefore, listed independently: WRES Local Server and WRES Web Service (not yet written).

With the side bar created, it opens up by default. What I'm not a fan of is that it hides the table of contents for the current wiki. I'm going to see if I can get the current wiki ToC to display below the above list, so that both the overall list of wikis and ToC for the current wiki are displayed.

Hank

**Comment by HankHerr-NOAA** [comment]
So, given the side bar is a separate markdown, I'm not sure that it can create a ToC for the displayed wiki markdown. I then investigate if the **Pages** dropdown can be modified, but I'm not finding anything there. If I can get the **Pages** to be open by default (along with the custom sidebar) and then constrain pages to only display the current page, it should default to show the ToC.

Also realized I've been a bit of a doofus all this time: textile is an option available in the GitHub wiki. We could have just copy-and-pasted wikis from VLab with changes to links and images and be done with it. I think. Unless markdown has some natural advantage over textile that would make it preferred. Since I've already migrated 25 pages to markdown, I'll just stick with markdown.

Hank

**Comment by james-d-brown** [comment]
Sticking with markdown makes sense, I think.

**Comment by HankHerr-NOAA** [comment]
So I can't find any options related to the wiki or the *Pages* default sidebar.  By including a custom side bar, we are hiding the *Pages* sidebar and hiding the ToC for the current page.  

In researching adding a ToC to the top, automatically, I can find tools that would be run from the command line, using a clone wres.wiki repo, but have not had luck finding one that would be embedded in the wiki, itself, requiring no command-line actions.  I think James or Evan did a bit of research above, so I'll see what they found. I know that HTML is acceptable in the markdown, so I may look at some JavaScript embedded in HTML or something like that. In fact, let me start by web searching that to see if anything pops up.

Hank

**Comment by james-d-brown** [comment]
Sounds like it may work with MediaWiki and __TOC__, but then we need to change from markdown, sigh. May be a battle not worth fighting, but it definitely makes the longer wikis MUCH harder to use, particularly the declaration language.

**Comment by HankHerr-NOAA** [comment]
It just bugs me that they create a really nice readme viewer that works with markdown, but don't use that viewer for the wiki.

Hank

**Comment by HankHerr-NOAA** [comment]
So, I think I'm done with my research here. I have some other work to do.

First, there appears to be no way to display the same ToC available in the default sidebar within a custom side bar on the right. At least, not that I can find with web searching. 

Regardless, I think a ToC at the top, with more width real-estate, would be better. However, an automatically generated ToC added to the wiki content, itself, does not seem to be an option. That leaves a manually constructed one which would be a pain to maintain.  However, it should be readily possible to download a tool or write a simple bash script that can take a markdown, parse out the headers, and arrange them into a table of contents at the top. The script could probably also be written to remove the existing ToC before adding the new one, as well. If I were to guess, I'd say that's probably 20-30 lines of bash, and might actually be preferred over linking in some other tool, depending on how simple those tools are.

Using that approach, the wiki author will update the wiki, save it, then update their wiki clone, run the script on that wiki page, and push the changes back to the wiki repo.

**If this is worth it, in your opinion, please let me know.** It would probably take a day or less to do the work.

Thanks,

Hank

**Comment by james-d-brown** [comment]
Main comment is that the TOC needs to be re-arranged a bit. I think most of the stuff under "declaration language" is not about declaration. Everything under there, except for the "Complete examples..." should be in a separate wiki.

Conversely, "Configuring the Raw NWM Data Source" sounds a lot like declaration and the content suggests that too, so I would put that under the declaration grouping and reword "Configuring" to "Declaring".

The first group also contains an odd mixture. I think we should break out all the wikis on input format requirements into a separate group, likewise for any output format requirements/descriptions. The introductory resources should be a separate thing too.

Overall, the current grouping is very shoehorned rather than thematic. When breaking out the new top-level groups, I would also ensure that the declaration language is the second grouping, after the instructions for using wres.

**Comment by HankHerr-NOAA** [comment]
I chose that order based on what links to what, not based on wiki content.  When there was a tie (i.e. both instructions and declaration language point to a wiki), I usually put it in the declaration language wiki if it was describing a feature. Point is, if there is a preferred logical structure, I can make that happen, just give me the top-level headings you would recommend and I'll restructure it.

Also, note that I don't want to own the wiki, so if you believe there is a better way to organize, feel free to make those changes. The side bar markdown is very small, so reordering it would take a couple/few minutes only.  If that. 

Hank

**Comment by HankHerr-NOAA** [comment]
> Conversely, "Configuring the Raw NWM Data Source" sounds a lot like declaration and the content suggests that too, so I would put that under the declaration grouping and reword "Configuring" to "Declaring".

I made the name change to the wiki. Again, since this wiki is only referenced in the instructions, and not in declaration language, I had positioned it under instructions.  If we want to group it with the declaration language, then we should reference it within that wiki somewhere, or just make the decision that the groupings on the side bar will not be related to how wikis are linked.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
I reordered the sidebar a bit. That's probably it for my changes today. 

Next week, I'll do any wrap-up work of the existing wikis that is identified, resolve this ticket, create new tickets for the none gaps (earlier comment), and then start connection the WRES User Support wiki to the GitHub wiki. Specifically, I'll want a single, top-level wiki with all of the COWRES-specific capabilities described (details in linked-to wikis where appropriate), and then I'll want any wikis that will remain in VLab (including the Cookbook) edited to point to GitHub for any wikis that were moved.

More next week,

Hank

**Comment by james-d-brown** [comment]
> > Conversely, "Configuring the Raw NWM Data Source" sounds a lot like declaration and the content suggests that too, so I would put that under the declaration grouping and reword "Configuring" to "Declaring".
> 
> I made the name change to the wiki. Again, since this wiki is only referenced in the instructions, and not in declaration language, I had positioned it under instructions. If we want to group it with the declaration language, then we should reference it within that wiki somewhere, or just make the decision that the groupings on the side bar will not be related to how wikis are linked.
> 
> Thanks,
> 
> Hank

I don't think they necessarily need to be linked. These groupings are simply thematic, for logical access via the TOC. But if there is an obvious place to link, that is fine too.

**Comment by james-d-brown** [comment]
Organization still needs work. The stuff under capabilities should probably come further down and it is probably "Miscellaneous capabilities". All top-line items should probably be links to wiki resources rather than a mix, so we may need to create some extra (short) wikis that provide a front door for those. It's getting better, though. Thanks for working on it.

**Comment by HankHerr-NOAA** [comment]
See https://github.com/NOAA-OWP/wres/wiki/Declaration-Language. I've created a script that adds a table of contents at the top of any markdown. I borrowed the script from another GitHub source without modification:

https://github.com/Lirt/markdown-toc-bash/blob/main/markdown-toc.sh

The full script is provided below; its bash, so there is almost certainly a more efficient way to get the job done. 

One hacky solution I had to come up with was how to mark the end of the table of contents, so that I could replace it with a new table of contents if run again. Rather than relying on white space (i.e., read the ToC until you reach a blank line) or something else that could easily be modified by accident, I added a comment in the markdown file, itself:

    

I can then remove everything before and including that marker when building and inserting a new ToC.  Anyway, the details are in the script, below, noting that the github script was renamed to, `generate_toc_from_md.sh` locally.  

My bigger question is this: **Do we want to use this script?** I did not get any feedback when I asked that before (unless I overlooked it). The process would be as follows:

1. The author writes a new wiki that requires a table of contents (subjective judgment). 
1. The author checks out the latest from the `wres.wiki` repo.
1. The author runs the script, `./scripts/add_toc.sh [filename]`.
1. The author reviews the change made by the script in the `[filename].new` file.
1. If its good, the author moves the `.new` file to the original file name, overwriting it, and then pushes the changed markdown.

We can eventually remove the review step and directly modify the wiki once we trust the process and script. 

Thoughts?

Hank

**Comment by james-d-brown** [comment]
It's a bit fugly, but the outcome is more important than the journey in this case, so I am fine with it. It is probably only needed for the longer wikis.

**Comment by HankHerr-NOAA** [comment]
I forgot to share the script. Since I can't figure out how to refer you to the script through a URL after I pushed it to the repo, I just copied-and-pasted it below.

I'm going to identify some other wikis that need a ToC and add it later today.

Hank

=========================================

```bash
#!/bin/bash

# Takes a single argument: the markdown file to which to add a table of contents.  

filename=$1
tmpfile=/tmp/temporary_file.txt
tmptoc=/tmp/temporary_toc.txt 
END_OF_TOC_MARKER="END OF TABLE OF CONTENTS MARKER, DO NOT DELETE"

echo "Processing file $filename..."

# If there is already a table of contents, then remove it and creat temp file.
grep -Fq "# Table of Contents" $filename
code=$?
if [ $code -eq 0 ]
then
    echo "Found an existing \"Table of Contents\"."

    # If there are at least two lines, create a temporary file by removing ToC
    if [[ $(grep '^#' $filename | wc -l) -ge 2 ]]
    then
        echo "There are at least two headings, proving that the table of contents is meaningful. Now running:"

        # Get the line number for the end of ToC marker and the last line of the file.
        number=$( grep -n "$END_OF_TOC_MARKER" $filename | cut -f1 -d: )
        if [ $? -ne 0 ] 
        then
            echo "A "Table of Contents" was found, but no end-marker, \"$END_OF_TOC_MARKER\"."
            echo "I'm not sure how to process this. Aborting."
            exit 1
        fi

        # Increment the line number by two to jump over the end marker and blank line
        # added in echo statements, below.
        number=$((number+2))
        lastline=$(wc -l < $filename)

        echo "Line number of end of ToC marker is $number. File has $lastline many lines."
        echo "Creating a temp file starting with the line after $number."

        # Copy the file including and after that line number to the temp file.
        sed -n "${number},${lastline}p" $filename > $tmpfile

    # Else, return an error.
    else
        echo "The file appears to only contain a table of contents. That's not good. Erroring out."
        exit 1
    fi
# Otherwise, just create the temp file using what is given.
else
    echo "No table of contents found; adding a new one at the top."
    cp $filename $tmpfile
fi

# Generate the ToC in a temporary file
./generate_toc_from_md.sh $filename > $tmptoc

cat $tmptoc > $filename.new
echo "<!--- $END_OF_TOC_MARKER -->" >> $filename.new
echo "" >> $filename.new
cat $tmpfile >> $filename.new

echo "Markdown with ToC has been created in $filename.new"


```

**Comment by james-d-brown** [comment]
One minor thing we could consider removing is the bullets (not the indentation) as they seem redundant with the enumeration, but that is minor.

**Comment by HankHerr-NOAA** [comment]
Right. But the enumeration isn't present for all of the wikis... I think. Actually, let me take a look at that, first.

If it's not present, then we could still add an option to control if bullets are used.

Hank

**Comment by james-d-brown** [comment]
If it isn't, we should probably aim for consistency across all wikis.

**Comment by james-d-brown** [comment]
In general, an enumeration is easier to refer to when directing users. For example, "see section N of the declaration language wiki."

**Comment by james-d-brown** [comment]
However, when linking internally within a wiki, it's best to remove the enumeration from the link name because the enumeration could change over time and updating internal link names creates friction.

**Comment by HankHerr-NOAA** [comment]
Agreed on consistency. This is isn't the only inconsistency either. I pointed out one a while ago where we are not consistent in our use of capitalization for wiki titles. 

As you said in your follow up comment, I think one reason we avoided enumeration was that it makes it harder to maintain.  

I've add a ToC to https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Using-WRES, as well, which includes a mix, with the instructions themselves enumerated, but sub-sections not.  

I'm going to investigate markdown options for adding a number to headers, but not requiring numbers for an external reference.  Our current enumeration technique (for pages that have it) is manual, so the number will be required when referring to section headers... to the best of my knowledge. 

Hank

**Comment by HankHerr-NOAA** [comment]
I'm not finding any reasonable options for auto-numbering headers in a wiki within GitHub. There are tools that can do it (such as Pandoc), but GitHub uses its own "Markdown" tool and see no examples indicating its possible to auto-number with that. As such, any internal reference to a numbered section in a wiki will need to include the number. That would also be true of external references. In general, as soon as the name of that header needs to change because of a change to the numbering, all referring links would need to be updated accordingly.

For this,

> In general, an enumeration is easier to refer to when directing users. For example, "see section N of the declaration language wiki."

when I refer to a wiki without enumeration, it would typically be something like, "see this section of the Declaration Language wiki", with "this section" linking to the section, and "Declaration Language" linking to the top-level. In other words, I avoid section numbers in most cases. But perhaps that makes the wiki less readable. 

Anyway, we'll probably want to discuss section numbering and other consistency issues a bit more during a dev call in the future. Thanks,

Hank

**Comment by james-d-brown** [comment]
Take a look at the declaration language wiki. All of those internal references within the page omit the enumeration for precisely this reason. The html tag is composed of the actual url and a name to display, the latter being completely arbitrary. It is a pain, but it makes it more maintainable, and we can continue that rule.

**Comment by HankHerr-NOAA** [comment]
I'll take a look at that wiki.

I've completed a first draft of the units of measurement wiki, https://github.com/NOAA-OWP/wres/wiki/Units-of-Measurement. I still need to proof read it.

Hank

**Comment by HankHerr-NOAA** [comment]
I may need more specific guidance on where to look in the Declaration Language wiki. 

Here is what I see in the VLab wiki's first sentence of Section 6.1, which I think contains one of the internal references to which you are referring:

     Geographic features may be declared explicitly by listing each feature to evaluate ("How do I declare a list of geographic features to evaluate?":https://vlab.noaa.gov/redmine/projects/wres-user-support/wiki/Declaration_language/#62-How-do-I-declare-a-list-of-geographic-features-to-evaluate).

In the migrated GitHub wiki, it looks like this (it becomes Section 4.2 because I removed Section 3 and 4 from VLab, since I don't think mentioning the old XML language is needed):

    Geographic features may be declared explicitly by listing each feature to evaluate ([How do I declare a list of geographic features to evaluate?](https://github.com/NOAA-OWP/wres/wiki/Declaration-Language#42-how-do-i-declare-a-list-of-geographic-features-to-evaluate)).

In both cases, VLab and GitHub, the section number shows up in the link without the '.' in the middle.  So the user doesn't see the section number in the internal references, which is something, but the linked URL will break if a section number changes. So, if a section is inserted resulting in new numbers for sections below the insert point, then every internal reference within the declaration language  to impacted sections will still need to be updated.  

So I'm not sure I'm looking at the right example. Thanks,

Hank

**Comment by james-d-brown** [comment]
For sure, the links will break, which is a PITA, but the labels won't break. I don't see an outcome that doesn't have at least one downside. The downside of not enumerating them is that they are harder to reference, especially in long wikis.

Either way, I think consistency across wikis probably trumps all so, if we're going with no enumeration, then it should be consistent.

**Comment by james-d-brown** [comment]
What is the worst outcome? Probably having broken wiki links? If true, then eliminating the enumeration everywhere is probably the way forward.

**Comment by HankHerr-NOAA** [comment]
> What is the worst outcome? Probably having broken wiki links? If true, then eliminating the enumeration everywhere is probably the way forward.

Honestly, I think the worst outcome is having the wiki be harder to read because there are no section numbers. If that's true, then we put the section numbers in and try to make updating links as painless as possible (again, your recommendation would yield one thing to change per section number change instead of two, which is something).  Note that, with access to the markdowns via git, we can readily use `grep` to find URLs that need changing, so it shouldn't be too painful. Hopefully.

I guess the question is, how much harder to read are the wikis without section numbers?  For some wikis, particularly smaller ones, it probably doesn't matter. However, for larger ones, like the Declaration Language wiki, the section numbers can help navigate it quickly. The again, we will have a ToC at the top to help with that.  Hmmm...

Something to discuss, I guess.

Hank

**Comment by james-d-brown** [comment]
I guess one advantage of broken links in github wikis is that they show up very clearly, so a quick scan would reveal any broken ones (at least while the wikis remain small enough in number and size to scan by eye).

Tough call. I agree that ditching numbers from the declaration language wiki will make it harder to navigate and reference.

**Comment by james-d-brown** [comment]
... but if we retain it in one place, we really need to do it everywhere. Consistency is right up the top of priorities.

**Comment by james-d-brown** [comment]
In fact, looking again at the declaration language wiki, it would be pretty awful to navigate without the numbers because the difference in heading font between layers is not sufficiently distinct.

**Comment by james-d-brown** [comment]
I assume we can't apply the same trick with distinct url / naming to the headers themselves, i.e., use html headers? That way, the link wouldn't break and the only thing that would need to be changed would be the heading label itself (the thing that minimally needs to be changed when adding new sections).

**Comment by HankHerr-NOAA** [comment]
Right. It definitely helps with Declaration Language; I use the numbers all of the time. Adding numbers to some of the smaller wikis just feels like overkill, but I understand the point about consistency. Might become a maintenance headache, however. 

My workday is done. I have some further editing of the units wiki to do tomorrow and then I'll look for more wikis that could be helped by a ToC and add them where its appropriate.

More tomorrow,

Hank

**Comment by HankHerr-NOAA** [comment]
The units of measurement wiki has been proofread and is ready for review:

https://github.com/NOAA-OWP/wres/wiki/Units-of-Measurement

I think this is a wiki where a ToC would be helpful (it would state the questions being answered up front), but enumerated sections would be unnecessary. So, if we strive for consistency and add enumeration, we would be adding more maintenance later for relatively little gain (imho).  

As mentioned before, I'd like to discuss this during the -dev call. Specifically, how important do we weigh consistent enumerated sections vs maintenance. I think that depends on how important consistent enumerated sections is to readability across wikis. More on that Thursday, I guess.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Pending more discussions on Thursday, I'm going to create a VLab ticket for an updated User Support wiki entry page that better dovetails with the GitHub wiki and start drafting it.  My intention is to write it on the side and only switchover once we are happy with it. 

Thanks,

Hank

**Comment by james-d-brown** [comment]
Fair enough. There is no perfect solution to this. But I do agree that, on balance, users will benefit from the enumeration of the very long wikis, like the declaration language wiki, more than the developers will suffer from the headache of maintaining them.

**Comment by james-d-brown** [comment]
Correct. Kinda solves all problems, but possibly at the expense of dirtier markdown (not exactly sure what the markdown standard says about embedded html, such as it is).

**Comment by HankHerr-NOAA** [comment]
I've done quite a bit of web searching which is bearing no fruit.  Any text written in the markdown header shows up in the link with some exceptions, like punctuation and upper case being converted to lower case. So I can't wrap a section number in HTML inside of a header title in the hopes that the number would then not appear in the link URL provided when viewing it.

In fact, when looking at the raw HTML for headers, `<h#>`, I don't see an option for specifying the link created for the header. You can have the header link to a custom URL using `href`, but that does not impact the link others have to use to link to that section.

I'm going to do a bit of research into anchors to see if that can be leveraged (I know next to nothing about HTML, so this is all new to me). But, if not, my only idea is to use a command-line tool to add section numbers to headers and maintain links to those headers accordingly.  

Hank

**Comment by HankHerr-NOAA** [comment]
So I think setting an anchor for each enumerated header in a markdown is an option.  For example, if I wanted to create a non-enumerated link for an enumerated header, it may look like this:

```
<a name="what-are-the-other-capabilities-of-the-wres"></a>
# 1.1 What are the other capabilities of the WRES?
```

That header could then be linked using either `...#what-are-the-other-capabilities-of-the-wres` or `...#11-what-are-the-other-capabilities-of-the-wres`, but the former will be preferred/required.  

For that to work, whenever anyone creates an enumerated header, they would need to manually add an anchor with the same name as the anchor for the header, but with the leading section number and hyphen removed.  See the example above. Other parts of the wiki would then link to that anchor.

I think this is better than having the link include a section number that can easily change, but its not an optimal solution. 

Thoughts?

Hank

**Comment by james-d-brown** [comment]
Agree, it is suboptimal, but probably the best among suboptimal things we've identified so far.

I would probably go with that. At a certain point, it becomes too much of a time sink to continue. I think that is a reasonable compromise for users and developers alike.

**Comment by HankHerr-NOAA** [comment]
I'll prototype the idea with Declaration Language. Note that other wikis are ]linking sections of the Declaration Language using the current headers. I'll either fix those on a case-by-case basis when problems arise or try to come up with some sort of slide shell commands to make the changes.

Hank

**Comment by HankHerr-NOAA** [comment]
The Declaration Language wiki has been updated. To test it out, go to the table of contents, click on a section, modify the URL in the browser to remove the leading numbers and hyphen, and press enter. The browser should remain on that section, but will shift up just slightly because the anchor is on the line preceding the section header.  I prefer the anchor being on the line just above, since on the same line at the end will result in very long header lines (if it even works; I didn't test it), and on the line after will result in the page displaying with the  header slightly cutoff.  

I also did some other cleaning of Declaration Language since it was one of the first migrated wikis. Specifically, brackets were migrated with an escape character, '\', which was unnecessary.  There is probably some other stuff that still needs cleaning, as well.

Hank

**Comment by james-d-brown** [comment]
Sounds good. 

So take this for example:

```When no `output_formats` are declared, the software will write the `csv2` format, by default. For example, when considering the simplest possible evaluation described in [What is the simplest possible evaluation I can declare?](https://github.com/NOAA-OWP/wres/wiki/Declaration-Language#1-what-is-the-simplest-possible-evaluation-i-can-declare), no `output_formats` are declared and `csv2` will be written.```

We should now be able to remove the reference to `#1-` in the link. That seems to work...

Are you planning to update all those internal enumerated URLs to the anchor URL?

**Comment by HankHerr-NOAA** [comment]
For Declaration Language, yes. In fact, the example you provide illustrates a mistake I made in the migrations for the first few wikis. Specifically, instead of using `[[...]]` notation, I explicitly defined the link. That's not necessary and makes for possible maintenance issues down the road.  

I believe the `[[...]]` notation should work for anchors, as well, but let me double check that.

Hank

**Comment by HankHerr-NOAA** [comment]
Oh well. Anchors do not appear to work inside of the `[[...]]` notation. So, if we are going to point to an enumerated section, we need to manually define the link using the `[...](...)` notation and refer to the anchor instead of the heading. Sigh.

I'm going to confirm that via web-search. If an alternative does not jump out at me, I'll make the changes to use the anchors in explicit URLs.

Hank

**Comment by james-d-brown** [comment]
No, I think we're in agreement about the enumeration. On balance, while consistency would be nice, we don't need to enumerate  a short wiki with a small number of sections. The issue arises with longer wikis, particularly those with different levels of heading.

**Comment by HankHerr-NOAA** [comment]
Oh, okay.  Well, may not be much to talk through, then.

Tomorrow (not enough time today), I'll look for other wikis that may need section numbers. The Declaration Language wiki is, by far, the longest. Perhaps the Instructions wiki needs them, as well. Currently, the section numbers are actually intended to be Step numbers (Step 1, Step 2, etc.), not sections, and that's why I never enumerated sub-sections within the steps.  I'll revisit that decision tomorrow.

I'll also add ToCs to the wikis that need them. In addition to the number of headers, I think a good way to judge if a ToC would be helpful is if the questions asked in the section headers would be best to summarize at the top to make them easier to find. Chances are, more wikis will have a ToC added than section numbers, but maybe I'll decide differently.

I also need to return to the draft top-level VLab wiki.

Documentation is so much fun,

Hank

**Comment by HankHerr-NOAA** [comment]
Noting here that I need to migrate this wiki to GitHub, or at least compare it with what is already available and decide if it needs migration:

https://vlab.noaa.gov/redmine/projects/wres-user-support/wiki/Obtaining_and_using_the_WRES_as_a_standalone_application

Hank

**Comment by HankHerr-NOAA** [comment]
The found the old "WRES_Geographic_Location_Table" wiki in VLab which is woefully out of date. 

The WRDS team has provided me a text file with a PSQL dump of the crosswalk that is up-to-date with the latest location service. The format looks like the snippet below.  

Is this something we want to share with our GitHub users or just the COWRES users in VLab? I'm actually leaning toward the former, since there may be folks who would like to have that crosswalk on hand, even if they can't use the COWRES. Thoughts?

The file uses pipe-delimiters currently; I'm guessing reformatting to a CSV would be best, right?

Thanks,

Hank

========================

```
 nws_station_id |     gage_id     | nwm_feature_id 
----------------+-----------------+----------------
 WRFC2          |                 |        2900123
 ICIW1          | 12458000        |       23081224
 IKPA2          | 15820000        |               
 DARA4          | 07258000        |        7790194
 BCNA1          | 02374745        |        2413254
                | 054279465       |       13294264
 GRBN4          | 01403600        |        9513154
 4S2            |                 |       24506070
 BUFN7          |                 |        9251294
                | 06022000        |         240380
 RAPN7          | 02080100        |       12050004
 LCHC2          | 06714500        |        2885314
                | 13094700        |       24490962
                | 02HC032         |       41027540
 MBHN7          | 0208524090      |        8777381
                | 07094900        |         922103
```

**Comment by james-d-brown** [comment]
Up to date, for now. The problem with putting this somewhere public is maintaining it. Why would you rely on a (probably out-of-date) manual crosswalk when you have access to the WRDS feature service? How many folks that don't have access would want this crosswalk? Second question is a genuine question - if there could be more than a handful, then I'd probably still wait for the first person that asks for it.

**Comment by HankHerr-NOAA** [comment]
Its that second question that's of interest to me, as well. 

For the first question, I think it could be informative for some folks that want to know the crosswalk that WRDS is going to use, perhaps confirming it for accuracy. However, that's only likely to be done if issues are encountered.  

I can go along with the recommendation that we want for someone asking for it. In the mean time, I'm going to attach the file to the existing, out of date VLab wiki, and then move that wiki to the "archive" bucket so we don't lose it.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Reiterating the GitHub documentation gaps, probably in order of priority:

1. The VLab wiki for obtaining and using a standalone needs to be migrated. Its a more informative wiki than what we've provided in the README.md and the top level wiki, so I think it would be good to migrate it and refer to it where appropriate, modifying the README.md and top level wiki if needed.

2. The WRES Web Service wiki needs to be written. This should target an advanced user that wants to setup their own cluster instance of the WRES web service. 

3. A developer wiki should be written, providing any information we believe developers will need to contribute to the WRES.

I believe item 1 should be done for this ticket. Items 2 and 3 can become their own tickets so that this ticket can be resolved once 1 is done. And once we have sufficient reviews. 

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
First, I'm going to migrate the standalone wiki now. I have other work to do, specifically for the new Postgres server in staging, but I want to get this out of the way.

Second, I'm leaning toward including the ToC in all of the wikis longer than one screen (approximately) and that have more than one section.  Most of the wikis I looked at yesterday need a ToC, since important sections headers (questions in most cases) can be off the screen requiring navigation to get to. Its import that folks see those section headers (questions) up top so that they know they exist. I don't know if I'll get to that today. It depends on how quickly I can make progress on the Postgres server.

Migrating the standalone wiki,

Hank

**Comment by HankHerr-NOAA** [comment]
Evan, James:

How do we validate a downloaded release .zip from https://github.com/NOAA-OWP/wres/releases? Where is the checksum? Does it still use `sha256sum`?

Hank

**Comment by james-d-brown** [comment]
Look under the assets for `sha256checksum.txt`. Last release (6.24), they were uploaded manually, next it will be automated.

**Comment by james-d-brown** [comment]
Aside, it would be nice if we could break the assets into two groups, the first containing the main release and the second group containing everything else. However, I'm not sure this is easy/possible. Reason: as an average user, it may be hard to find the actual thing they need, even though we do provide instructions on what to download.

**Comment by HankHerr-NOAA** [comment]
Agree. The list of zips is long, and the likelihood of a wrong-click or a miss-click are high. But, as you said, it may not be easy or possible.

The standlone wiki is prepared:

https://github.com/NOAA-OWP/wres/wiki/Obtaining-and-using-the-WRES-as-a-standalone-application

The reason it slipped through the cracks is because we only exposed that wiki via a News item. We did not link it through the main wiki, since we did not want to encourage its use. With GitHub, I think the documentation needs to be directly linked. So I modified the wiki appropriately (primarily the download instructions and removing the section numbers which I didn't think were necessary) and dovetailed it into the main and other wikis.  

Hearing no concerns about my plans for the ToC, I'm going to pull the wikis, update those that need a ToC, and push.  

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
ToCs are in place for the wikis that I judged to need them, which was most. 

I did spot some cleanup work that needs to be done for the earlier wikis I migrated. Recall that the early wikis were migrated with the tool auto-line wrapping. I figured out how to turn that off for newer wikis, and I think it is a big enough improvement to back-correct those early wikis. 

For now, however, I need to turn my focus to the new Postgres server.

Hank

**Comment by HankHerr-NOAA** [comment]
Developer wiki ticket is #57 and the web-service wiki ticket is #56 . 

Only work that remains is an hour or two of cleanup for the first wikis I migrated. I have that on my calendar to complete tomorrow. Once done, that will wrap up this ticket... pending reviews.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Also, there are some anchors that need to be put in place, but I was planning to do that when the need arose in the future. All anchors in the Declaration Language wiki are in place, but the instructions and some other wiki's may need them as well. If someone really wants them put in place now, let me know.

More tomorrow,

Hank

**Comment by james-d-brown** [comment]
Only in the sense that there's a better than even chance it will never happen if not done now. But I think their absence from the declaration language wiki would be the biggest pain, by far, so that is covered.

**Comment by HankHerr-NOAA** [comment]
Cleaning up the earlier migrated wikis... One of the main changes is to remove the new-line line wrapping. Apparently, the tool I identified for migrating from textile to markdown can automatically remove them by migrating from markdown to markdown and selecting "No wrap".  It will interpret where the new-lines are only reflecting a line wrap and remove them. That will make this much easier.

Hank

**Comment by HankHerr-NOAA** [comment]
I'm done with my clean up and reviews.  I'm going to mark this ticket as done and assign it to UAT, where UAT will be a review of the documentation, not necessary thorough, and a thumbs up during the 6.25 deployment.

Hank

**Comment by HankHerr-NOAA** [comment]
Marking done,

Hank

**Comment by ArvinEsmailzadeh-NOAA** [comment]
Reviewed this wiki: https://github.com/NOAA-OWP/wres/wiki/Obtaining-and-using-the-WRES-as-a-standalone-application

Looks Good.

**Comment by HankHerr-NOAA** [comment]
Note this comment:

https://github.com/NOAA-OWP/wres/issues/20#issuecomment-2246125348

Where can I document the two items noted in that comment for future reference? It might belong in a developer wiki eventually.

Hank

**Comment by james-d-brown** [comment]
Can you create a simple dev wiki now with access for users with repo write access only and place it there? We will need to migrate the dev wikis from vlab eventually, but we can start with a simple page to accumulate stuff that doesn't fit on the vlab dev wiki (like this example).

**Comment by HankHerr-NOAA** [comment]
Sure. It will be a one section wiki with an "under construction" header of some sort, but I can do that today. 

Hank

**Comment by HankHerr-NOAA** [comment]
Wiki created:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Developers

We can edit it further later. There is almost certainly more to say about editing wikis.

Hank
